src/cbmc_starter_kit/ctagst.py

```diff
@@ -0,0 +1,158 @@
+"""Ctags support for locating symbol definitions"""
+
+from pathlib import Path
+import json
+import logging
+import subprocess
+import sys
+
+################################################################
+# This popen method is used to subprocess-out the invocation of ctags.
+# This method duplicates code in other modules to make this ctags
+# module a stand-alone module that can be copied and reused in other
+# projects.
+
+def popen(cmd, cwd=None, stdin=None, encoding=None):
+    """Run a command with string stdin on stdin, return stdout and stderr."""
+
+    cmd = [str(word) for word in cmd]
+    kwds = {'cwd': cwd,
+            'text': True,
+            'stdin': subprocess.PIPE,
+            'stdout': subprocess.PIPE,
+            'stderr': subprocess.PIPE}
+    if sys.version_info >= (3, 6): # encoding is new in Python 3.6
+        kwds['encoding'] = encoding or 'utf-8'
+    try:
+        logging.debug('Popen command: "%s"', ' '.join(cmd))
+        logging.debug('Popen stdin: "%s"', stdin)
+        with subprocess.Popen(cmd, **kwds) as pipe:
+            stdout, stderr = pipe.communicate(input=stdin)
+        logging.debug('Popen stdout: "%s"', stdout)
+        logging.debug('Popen stderr: "%s"', stderr)
+        if pipe.returncode:
+            logging.debug('Popen command failed: "%s"', ' '.join(cmd))
+            logging.debug('Popen return code: "%s"', pipe.returncode)
+            raise UserWarning(f"Failed to run command: {' '.join(cmd)}")
+        return stdout, stderr
+    except FileNotFoundError as error:
+        logging.debug("FileNotFoundError: command '%s': %s", ' '.join(cmd), error)
+        raise UserWarning(f"Failed to run command: {' '.join(cmd)}") from error
+
+################################################################
+
+def ctags(root, files):
+    """List symbols defined in files under root."""
+
+    root = Path(root)
+    files = [str(file_) for file_ in files]
+    return (universal_ctags(root, files) or
+            exhuberant_ctags(root, files) or
+            legacy_ctags(root, files) or
+            [])
+
+################################################################
+
+def universal_ctags(root, files):
+    """Use universal ctags to list symbols defined in files under root."""
+
+    # See universal ctags man page at https://docs.ctags.io/en/latest/man/ctags.1.html
+    cmd = [
+        'ctags',
+        '-L', '-', # read files from standard input, one file per line
+        '-f', '-', # write tags to standard output, one tag per line
+        '--output-format=json', # each tag is a one-line json blob
+        '--fields=FNnK' # json blob is {"name": symbol, "path": file, "line": line, "kind": kind}
+    ]
+    try:
+        logging.info("Running universal ctags")
+        stdout, _ = popen(cmd, cwd=root, stdin='\n'.join(files))
+        strings = stdout.splitlines()
+    except UserWarning:
+        logging.info("Universal ctags failed")
```

Comment:
Suggest print out stderr when this happens.

With this implementation of popen, users currently need to rerun with `--debug` to get the actual error message from ctags. (and they need to know to run with `--debug`).

Reply:
No, the ctags() method invokes ctags as universal ctags, then exhuberant ctags, then legacy ctags, and uses the result of the first successful invocation (if any).   I don't want error messages printed during the fail down.  You might ask me to check the version and invoke the right version, but universal ctags can be built in many configurations and if the nice invocation of universal fails because it is missing the json extension, for example, I want to invoke using the exhuberant arguments that are reverse compatible.